PORT-11047 Create SSA for ADO pipeline copy and creation (#1684)
port-labs/port-docs: docs/guides/all/copy-pipeline-template-to-target-repo.md

---
sidebar_position: 2
displayed_sidebar: null
description: Learn how to copy pipeline template from one repository to another using Port Actions.
---

> # Copy Pipeline Template to Target Repository

Reviewer: ```suggestion
# Copy ADO Pipeline Template to Target Repo
```

This guide demonstrates how to copy a `pipeline.yaml` template from a base Azure DevOps repository to a target repository using a Port action and an Azure DevOps pipeline.

## Prerequisites
- Ensure you have a Port account and have completed the [onboarding process](https://docs.getport.io/quickstart).

> - The `Azure DevOps Repository` blueprint should be created during the [azure devops installation](/build-your-software-catalog/sync-data-to-catalog/git/azure-devops/#installation) process

Reviewer 2: Can they somehow have an alternative instead of them using the ADO ocean integration?

Reviewer: please make this clearer in the prerequisite. Write that the integration can be installed, or alternatively only the blueprint&mapping from there are enough.

> ## Implementation - copy pipeline.yaml template from one repository to another

Reviewer 2: title is long, let's make it short

Author: oaky

Follow these steps to get started :
1. Create an Azure DevOps Repository called `pipeline_copier` in your Azure Devops Organization/Project and [configure a service connection](/actions-and-automations/setup-backend/azure-pipeline#define-incoming-webhook-in-azure).

> 

Reviewer 2: can they use an existing repository for this guide?

Author: yeah they just have to create the azure-pipeline.yaml in that repo and configure service connection that's all

> :::note

Reviewer: refrain from using :::note pretty much everywhere, it doesnt stand out and its easy to miss. 
Also, make sure all admonitions have titles (there another note below, same comments for that one).

        "format": "url",
        "type": "string",
        "icon": "Link"
      },
      "readme": {
        "title": "README",
        "type": "string",
        "format": "markdown",
        "icon": "Book"
      },
      "defaultBranch": {
        "title": "Default Branch",
        "type": "string"
      }
    },
    "required": []
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {
    "project": {
      "title": "Project",
      "target": "project",
      "required": true,
      "many": false
    }
  }
}
```
</details>

<details>
  <summary>Azure DevOps repository mapping config</summary>

```yaml showLineNumbers
        - kind: repository
        selector:
          query: 'true'
        port:
          entity:
            mappings:
              identifier: .project.name + "/" + .name | gsub(" "; "")
              title: .name
              blueprint: '"azureDevopsRepository"'
              properties:
                url: .url
                readme: file://README.md
                defaultBranch: .defaultBranch  # Add this line
              relations:
                project: .project.id | gsub(" "; "")

> ```

Reviewer 2: if they don't use the ADO ocean integration then this will not work

Author: well in the SSA we are taking inputs from the repo blueprint. if they are not using the ocean integration, they have to  use API to ingest the data into port, use that blueprint for the frontend and select the target and base repo.

Reviewer 2: let's add a note specifying this alternative

</details>

> 

Reviewer 2: a lot of empty spaces

<br/>

3. Create Port action using the following JSON definition:

:::note

> Make sure to replace the placeholder for AZURE_DEVOPS_ORGANIZATION_NAME to where your `pipeline_copier`  repository resides.

Reviewer 2: in case they use an existing repository then we need to reword this `pipeline_copier`

Author: In the case of using an existing repository they don't need to create a new repo. so that admonition doesn't affect them. we can state that if you are using an existing repository just copy the azure-pipeline.yaml and put that in your existing repository. and do the necessary service connection configuration

Also validate that `invocationMethod.webhook` equals `port_trigger`.
:::

<details>
  <summary>Port Action</summary>

```json showLineNumbers
{
  "identifier": "copy_pipeline_template",

>   "title": "Copy Pipeline Template to Target Repo",

Reviewer 2: Create Azure Pipeline YAML from a Template?

Author: We take a pipeline template from a source repo, copy its content and recreate it in a target repository. Unless I didn't get the requirements right, that would be the way

  "icon": "Azure",
  "trigger": {
    "type": "self-service",

>     "operation": "DAY-2",

Reviewer 2: why is it a DAY-2 instead of CREATE?

Author: Well the target repo is an already existing repository.  so DAY-2 is modifying it to add a pipeline template. 
Unless they want the target repo to be created for them with the pipeline.yaml then the operation should be a DAY-2

          "icon": "Azure",
          "blueprint": "azureDevopsRepository",
          "format": "entity"
        }
      },
      "required": [
        "base_repo",
        "target_repo"
      ],
      "order": [
        "base_repo",
        "target_repo"
      ]
    }
  },
  "invocationMethod": {
    "type": "AZURE_DEVOPS",
    "webhook": "port_trigger",
    "org": "<AZURE_DEVOPS_ORGANIZATION_NAME>",
    "payload": {
      "base_repo_url": "{{ .inputs.base_repo.properties.url }}",
      "target_repo_url": "{{ .inputs.target_repo.properties.url }}",
      "base_repo_branch": "{{ .inputs.base_repo.properties.defaultBranch }}",
      "target_repo_branch": "{{ .inputs.target_repo.properties.defaultBranch }}",
      "azure_organization": "<AZURE_DEVOPS_ORGANIZATION_NAME>",
      "pipeline_file_name": "pipeline.yaml",
      "port_context": {
        "runId": "{{ .run.id }}"
      }
    }
  },
  "requiredApproval": false,
  "publish": true
}
```

</details>

:::note Organization and pipeline name
Replace `<AZURE_DEVOPS_ORGANIZATION_NAME>` with your Azure DevOps organization name. and `pipeline.yaml` with the name of the pipeline file you want to copy.

> :::

Reviewer 2: I saw you have mentioned this already above. so duplication?

Author: it's an oversight then

                -H 'Authorization: Bearer $(accessToken)' \
                -d '{"status":"SUCCESS","statusLabel":"Successfully copied file","message": {"run_status": "Copying finished successfully!" }}' \
                "https://api.getport.io/v1/actions/runs/${{ variables.RUN_ID }}"
            displayName: "Update Port with Success Status"

  - stage: update_run_status_failed
    dependsOn:
      - fetch_port_access_token
      - copy_and_create_pipeline
    condition: failed()
    jobs:
      - job: update_run_status_failed
        variables:
          accessToken: $[ stageDependencies.fetch_port_access_token.fetch_port_access_token.outputs['getToken.accessToken'] ]
        steps:
          - script: |
              curl -X PATCH \
                -H 'Content-Type: application/json' \
                -H 'Authorization: Bearer $(accessToken)' \
                -d '{"status":"FAILURE","statusLabel":"Failed to copy file","message": {"run_status": "Copying pipeline failed, check the logs in the pipeline" }}' \
                "https://api.getport.io/v1/actions/runs/${{ variables.RUN_ID }}"
            displayName: "Update Port with Failure Status"

```

</details>
<br/>

> 5. To configure the Pipeline in your project go to Pipelines -> Create Pipeline -> Azure Repos Git and choose `pipeline_copier` and click Save(in "Run" dropdown menu).

Reviewer: ```suggestion
5. To configure the Pipeline in your project go to Pipelines -> Create Pipeline -> Azure Repos Git and choose `pipeline_copier` and click Save (in "Run" dropdown menu).
```

   <br/>

6. Create the following variables as [Secret Variables](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/set-secret-variables?view=azure-devops&tabs=yaml%2Cbash):

    1. `PERSONAL_ACCESS_TOKEN` - a base64-encoded [Personal Access Token](https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows) with the following scopes:
        - Code: Full.
        - Release: Read, write & execute.

    2. `PORT_CLIENT_ID` - Port Client ID [learn more](/build-your-software-catalog/custom-integration/api/#get-api-token).
    3. `PORT_CLIENT_SECRET` - Port Client Secret [learn more](/build-your-software-catalog/custom-integration/api/#get-api-token).

<br/>

> 7. Trigger the action from the [Self-service](https://app.getport.io/self-serve) tab of your Port application.

Reviewer: ```suggestion
7. Trigger the action from the [Self-service](https://app.getport.io/self-serve) page of your Port application.
```

        steps:
          - script: |
              curl -X PATCH \
                -H 'Content-Type: application/json' \
                -H 'Authorization: Bearer $(accessToken)' \
                -d '{"status":"FAILURE","statusLabel":"Failed to copy file","message": {"run_status": "Copying pipeline failed, check the logs in the pipeline" }}' \
                "https://api.getport.io/v1/actions/runs/${{ variables.RUN_ID }}"
            displayName: "Update Port with Failure Status"

```

</details>
<br/>

5. To configure the Pipeline in your project go to Pipelines -> Create Pipeline -> Azure Repos Git and choose `pipeline_copier` and click Save(in "Run" dropdown menu).
   <br/>

6. Create the following variables as [Secret Variables](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/set-secret-variables?view=azure-devops&tabs=yaml%2Cbash):

    1. `PERSONAL_ACCESS_TOKEN` - a base64-encoded [Personal Access Token](https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows) with the following scopes:
        - Code: Full.
        - Release: Read, write & execute.

> 

Reviewer 2: are these enough to run the whole script?

Reviewer 2: Also users are having issue with passing the base64 encoded token. Can we ask them to pass the RAW token while we do the base64 ourselves in the script?

Author: I take the raw value and encode it in the pipeline, for the permission we need full access to code and some GitHub. I will add it

---
displayed_sidebar: null
description: Learn how to copy pipeline template from one repository to another using Port Actions.
---

# Copy Pipeline Template to Target Repo

> This guide demonstrates how to copy pipeline templates between Azure DevOps repositories using Port Actions. 
> 
> Once implemented:
> - Platform teams can define standard pipeline templates in a base repository
> - Developers can easily copy these templates to their repositories using a self-service action
> - Teams maintain consistent CI/CD configurations across projects

Reviewer: ```suggestion
This guide demonstrates how to copy pipeline templates between Azure DevOps repositories using a self-service action in Port. 

Once implemented:
- Platform teams can define standard pipeline templates in a base repository.
- Developers can easily copy these templates to their repositories using a self-service action in their portal.
- Teams can maintain consistent CI/CD configurations across projects.
```

> ## Common use cases

Reviewer: like the last guide we talked about, leave only one section for use-cases

- Standardize CI/CD configurations across multiple repositories
- Quickly set up new repositories with approved pipeline templates
- Ensure compliance with organizational build and deployment standards
- Enable developers to self-serve pipeline configurations within controlled boundaries

## Prerequisites

- Ensure you have a Port account and have completed the [onboarding process](https://docs.getport.io/quickstart).
- You can either:
    - Install the [Azure DevOps integration](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/azure-devops/#installation) to create the blueprint and mappings automatically, or
    - Alternatively, create only the `Azure DevOps Repository` blueprint and ingest repositories directly using [Port’s APIs](https://docs.getport.io/api-reference/create-an-entity)

## Set up infrastructure

First, let's set up the necessary Azure DevOps components to handle the pipeline copying process.

> ### Create pipeline copier repository

Reviewer: ```suggestion
### Create a pipeline copier repository
```

1. Create an Azure DevOps repository called `pipeline_copier` in your Azure DevOps Organization/Project and [configure a service connection](/actions-and-automations/setup-backend/azure-pipeline#define-incoming-webhook-in-azure).

> :::note Port trigger

Reviewer: this should not be in a callout

Use `port_trigger` for both `WebHook Name` and `Service connection name` when configuring your [Service Connection](https://learn.microsoft.com/en-us/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml)
:::

> :::info Option to use an existing repository

Reviewer: indent this callout

> :::info Option to use an existing repository

Reviewer: ```suggestion
:::info Use an existing repository
```

You may use an existing repository instead of creating a new one. Just ensure that you add the `azure-pipelines.yaml` file in **Step 4**  to the repository's root.
:::

> <br/>

Reviewer: ```suggestion
```

2. Update the `azureDevopsRepository` blueprint and mapping configuration with the `defaultBranch` property, depending on how your setup was created:

    - **If you installed the Azure DevOps integration**, update the `defaultBranch` property in the mapping config file.
    - **If you created the blueprint manually** (without the integration), add the JSON blueprint below and use Port's API to ingest the repository data.

> :::info Using Port's API 

Reviewer: include a link to the relevant API docs page

> 2. Update the `azureDevopsRepository` blueprint and mapping configuration with the `defaultBranch` property, depending on how your setup was created:
> 
>     - **If you installed the Azure DevOps integration**, update the `defaultBranch` property in the mapping config file.
>     - **If you created the blueprint manually** (without the integration), add the JSON blueprint below and use Port's API to ingest the repository data.
> 
> :::info Using Port's API 
> If you’re not using the Azure DevOps integration, you will need to use Port's API to ingest repository data based on the blueprint you created.
> :::

Reviewer: ```suggestion
2. Update the `azureDevopsRepository` blueprint and mapping configuration with the `defaultBranch` property, depending on how your setup was created:

    - If you installed the **Azure DevOps integration**, update the `defaultBranch` property in the mapping config file.
    - If you created the blueprint **manually** (without the integration), add the JSON blueprint below and use Port's API to ingest the repository data.

:::info Using Port's API 
If you’re not using the Azure DevOps integration, you will need to use Port's API to ingest repository data based on the blueprint you created.
:::
```

> Here’s the blueprint JSON configuration to use for manual creation:

Reviewer: ```suggestion
To create the necessary data model manually, use the following blueprint JSON and mapping configuration:
```

<details>
  <summary>Azure DevOps repository blueprint</summary>

```json showLineNumbers
{
  "identifier": "azureDevopsRepository",
  "title": "Repository",
  "icon": "AzureDevops",
  "schema": {
    "properties": {
      "url": {
        "title": "URL",
        "format": "url",
        "type": "string",
        "icon": "Link"
      },
      "readme": {
        "title": "README",
        "type": "string",
        "format": "markdown",
        "icon": "Book"
      },
      "defaultBranch": {
        "title": "Default Branch",
        "type": "string"
      }
    },
    "required": []
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {
    "project": {
      "title": "Project",
      "target": "project",
      "required": true,
      "many": false
    }
  }
}
```
</details>

<details>
  <summary>Azure DevOps repository mapping config</summary>

> ```yaml showLineNumbers

Reviewer: indentation is wrong

        - kind: repository
        selector:
          query: 'true'
        port:
          entity:
            mappings:
              identifier: .project.name + "/" + .name | gsub(" "; "")
              title: .name
              blueprint: '"azureDevopsRepository"'
              properties:
                url: .url
                readme: file://README.md
                defaultBranch: .defaultBranch  # Add this line
              relations:
                project: .project.id | gsub(" "; "")
```
</details>

<br/>

> 3. Create Port action using the following JSON definition:

Reviewer: ```suggestion
3. Create a self-service action in Port using the following JSON definition:
```

> :::tip Organization and repository placeholders

Reviewer: this callout is not clear, rewrite it in more accurate and clear language

          "blueprint": "azureDevopsRepository",
          "format": "entity"
        },
        "target_repo": {
          "type": "string",
          "title": "Target Repository",
          "icon": "Azure",
          "blueprint": "azureDevopsRepository",
          "format": "entity"
        }
      },
      "required": [
        "base_repo",
        "target_repo"
      ],
      "order": [
        "base_repo",
        "target_repo"
      ]
    }
  },
  "invocationMethod": {
    "type": "AZURE_DEVOPS",
    "webhook": "port_trigger",
    "org": "<AZURE_DEVOPS_ORGANIZATION_NAME>",
    "payload": {
      "base_repo_url": "{{ .inputs.base_repo.properties.url }}",
      "target_repo_url": "{{ .inputs.target_repo.properties.url }}",
      "base_repo_branch": "{{ .inputs.base_repo.properties.defaultBranch }}",
      "target_repo_branch": "{{ .inputs.target_repo.properties.defaultBranch }}",
      "azure_organization": "<AZURE_DEVOPS_ORGANIZATION_NAME>",
      "pipeline_file_name": "pipeline.yaml", # Update this if your pipeline file name is different
      "port_context": {
        "runId": "{{ .run.id }}"
      }
    }
  },
  "requiredApproval": false
}
```

</details>

> <br/>

Reviewer: ```suggestion
```

  - stage: update_run_status_failed
    dependsOn:
      - fetch_port_access_token
      - copy_and_create_pipeline
    condition: failed()
    jobs:
      - job: update_run_status_failed
        variables:
          accessToken: $[ stageDependencies.fetch_port_access_token.fetch_port_access_token.outputs['getToken.accessToken'] ]
        steps:
          - script: |
              curl -X PATCH \
                -H 'Content-Type: application/json' \
                -H 'Authorization: Bearer $(accessToken)' \
                -d '{"status":"FAILURE","statusLabel":"Failed to copy file","message": {"run_status": "Copying pipeline failed" }}' \
                "https://api.getport.io/v1/actions/runs/${{ variables.RUN_ID }}"
            displayName: "Update Port with Failure Status"

```

</details>
<br/>

5. To configure the Pipeline in your project go to Pipelines -> Create Pipeline -> Azure Repos Git and choose `pipeline_copier` and click Save (in "Run" dropdown menu). 

<br/>

6. Create the following variables as [Secret Variables](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/set-secret-variables?view=azure-devops&tabs=yaml%2Cbash):

>     1. `PERSONAL_ACCESS_TOKEN` - a [Personal Access Token](https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows) with the following scopes:

Reviewer: refrain from using numbered lists inside other numbered lists
